realm/realm-library/src/main/java/io/realm/internal/OsSet.java

```diff
@@ -90,15 +106,99 @@ public boolean remove(@Nullable Object value) {
+            String[] values = collection.toArray(new String[objects.length]);
+            return nativeContainsAllString(nativePtr, values);
+        } else {
+            // TODO: add missing types
+            throw new UnsupportedOperationException("set containsAll - Hold your horses cowboy...");
+        }
+    }
+
+    public boolean union(OsSet otherRealmSet) {
+        return nativeUnion(nativePtr, otherRealmSet.getNativePtr());
+    }
+
+    public <E> boolean addAll(Collection<? extends E> collection, Class<?> valueClass) {
+        return collectionFunnel(collection, valueClass, ExternalCollectionOperation.ADD_ALL);
+    }
+
+    public boolean asymmetricDifference(OsSet otherSet) {
+        return nativeAsymmetricDifference(nativePtr, otherSet.getNativePtr());
+    }
+
+    public <E> boolean removeAll(Collection<? extends E> collection, Class<?> valueClass) {
+        return collectionFunnel(collection, valueClass, ExternalCollectionOperation.REMOVE_ALL);
+    }
+
+    public boolean intersect(OsSet otherSet) {
+        return nativeIntersect(nativePtr, otherSet.getNativePtr());
+    }
+
+    public <E> boolean retainAll(Collection<? extends E> collection, Class<?> valueClass) {
+        return collectionFunnel(collection, valueClass, ExternalCollectionOperation.RETAIN_ALL);
     }
 
     public void clear() {
         nativeClear(nativePtr);
     }
 
+    public OsSet freeze(OsSharedRealm frozenSharedRealm) {
+        long frozenNativePtr = nativeFreeze(this.nativePtr, frozenSharedRealm.getNativePtr());
+        return new OsSet(frozenSharedRealm, frozenNativePtr);
+    }
+
+    private <E> boolean collectionFunnel(Collection<? extends E> collection,
+                                         Class<?> valueClass,
+                                         ExternalCollectionOperation operation) {
+        // Return unchanged if collection is empty
+        if (collection.size() == 0) {
+            return false;
+        }
+
+        // Collection cannot be removed if it is not the same type
+        Object[] objects = collection.toArray();
+        Class<?> itemClass = objects[0].getClass();
+        if (itemClass != valueClass) {
+            throw new IllegalArgumentException("Invalid collection type. Set and collection must contain the same type of elements.");
+        }
+
+        if (itemClass == String.class) {
+            String[] values = collection.toArray(new String[objects.length]);
+            switch (operation) {
+                case ADD_ALL:
+                    return nativeAddAllString(nativePtr, values);
+                case REMOVE_ALL:
+                    return nativeRemoveAllString(nativePtr, values);
+                case RETAIN_ALL:
+                    return nativeRetainAllString(nativePtr, values);
+                default:
+                    throw new IllegalStateException("Unexpected value: " + operation);
+            }
+        } else {
+            // TODO: add missing types
```

Comment:
How do you ensure that none of the TODO's are forgotten?

Reply:
We have an enum that exhaustively iterates over all supported set types in our tests. If a particular datatype hasn't been added here our tests will fail, and as you can see in Jira, there are tickets for all supported datatypes, so we (hopefully) we will not forget.